core/src/main/java/com/linecorp/armeria/common/stream/InputStreamStreamMessage.java

```diff
@@ -0,0 +1,362 @@
+    }
+
+    @Override
+    public boolean isOpen() {
+        return !completionFuture.isDone();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        if (isOpen()) {
+            return false;
+        }
+        final InputStreamSubscription inputStreamSubscription = this.inputStreamSubscription;
+        return inputStreamSubscription == null || inputStreamSubscription.position == 0;
+    }
+
+    @Override
+    public long demand() {
+        final InputStreamSubscription inputStreamSubscription = this.inputStreamSubscription;
+        if (inputStreamSubscription != null) {
+            return inputStreamSubscription.requested;
+        } else {
+            return 0;
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return completionFuture;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        if (!subscribedUpdater.compareAndSet(this, 0, 1)) {
+            subscriber.onSubscribe(NoopSubscription.get());
+            subscriber.onError(new IllegalStateException("Only single subscriber is allowed!"));
+            return;
+        }
+
+        if (executor.inEventLoop()) {
+            subscribe0(subscriber, executor, options);
+        } else {
+            executor.execute(() -> subscribe0(subscriber, executor, options));
+        }
+    }
+
+    private void subscribe0(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                            SubscriptionOption... options) {
+        final ExecutorService blockingTaskExecutor;
+        if (this.blockingTaskExecutor != null) {
+            blockingTaskExecutor = this.blockingTaskExecutor;
+        } else {
+            final ServiceRequestContext serviceRequestContext = ServiceRequestContext.currentOrNull();
+            if (serviceRequestContext != null) {
+                blockingTaskExecutor = serviceRequestContext.blockingTaskExecutor();
+            } else {
+                blockingTaskExecutor = CommonPools.blockingTaskExecutor();
+            }
+        }
+
+        final InputStreamSubscription inputStreamSubscription = new InputStreamSubscription(
+                subscriber, executor, blockingTaskExecutor, bufferSize, offset, length,
+                containsNotifyCancellation(options));
+        this.inputStreamSubscription = inputStreamSubscription;
+        subscriber.onSubscribe(inputStreamSubscription);
+    }
+
+    @Override
+    public void abort() {
+        abort(AbortedStreamException.get());
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+
+        final InputStreamSubscription inputStreamSubscription = this.inputStreamSubscription;
+        if (inputStreamSubscription != null) {
+            inputStreamSubscription.close(cause);
+        }
+        completionFuture.completeExceptionally(cause);
```

Comment:
It is likely that `inputStreamSubscription.close(cause)` could not be invoked if `abort()` is called first and `subscribe()` is called later. Honestly, it is a bug in `PathStreamMessage.abort()`.

Should we exceptionally complete `completionFuture` first and read `inputStreamSubscription` later?
```suggestion
        // `completionFuture` should be set before `inputStreamSubscription` is read
        // to guarantee the visibility of the abortion `cause` after
        // inputStreamSubscription is set in `subscriber0()`.
        completionFuture.completeExceptionally(cause);
        final InputStreamSubscription inputStreamSubscription = this.inputStreamSubscription;
        if (inputStreamSubscription != null) {
            inputStreamSubscription.close(cause);
        }
```
We also need to check if `completionFuture` is exceptionally completed after assigning `inputStreamSubscription`.
```java
final InputStreamSubscription inputStreamSubscription = new InputStreamSubscription(...);
this.inputStreamSubscription = inputStreamSubscription;
if (completionFuture.isCompletedExceptionall()) {
   // abort the subscriber with the cause of completionFuture.
   ...
}
```

Reply:
✅ updated! thanks for catching~! it's hard to consider this timing issue 😅